Added database name in the ResultSummary.

zhenlineo opened the pull request:
This requires the server to stream `db` with database name in the result summary.

Commit:
Added database name in the ResultSummary.

Commit:
Default to null when database name is absent for server versions lower than 4.0

ali-ince commented on driver/src/main/java/org/neo4j/driver/internal/util/MetadataExtractor.java:
```diff
@@ -99,12 +102,26 @@ public long extractResultAvailableAfter( Map<String,Value> metadata )
+    public static DatabaseInfo extractDatabaseInfo( Map<String,Value> metadata )
+    {
+        Value dbValue = metadata.get( "db" );
+        if ( dbValue == null || dbValue.isNull() || !dbValue.hasType( TYPE_SYSTEM.STRING() ) )
```
Shouldn't we throw `ProtocolException` if `!dbValue.hasType( TYPE_SYSTEM.STRING() )` holds?

ali-ince reviewed (approved):
looks good. I only think that we should throw `ProtocolException` on unexpected types in summary.

Commit:
Fix after review

Commit:
Unified the input and output value of database name.

The database name default to `null` if the server does not support multi-databases when the database name is returned in result summary.
However `null` is a forbidden value for database in session parameters. Then for user who want to connect to default db but do not want ot omit invoke `withDatabase` method on session parameter, they have to use empty string to work aroud our API design.

This PR allows `null` as a valid database name for session parameters, while disallows empty string as it is an invalid database name to users.
